src/core/cancellation_middleware.py

```
@@ -42,7 +46,9 @@ async def on_call_tool(self, context: MiddlewareContext, call_next):
             
         except Exception as e:
             duration_ms = (time.perf_counter() - start_time) * 1000
-            logger.error(f"Tool call failed after {duration_ms:.2f}ms: {e}")
+            logger.error(f"❌ CANCELLATION MIDDLEWARE: Tool call failed after {duration_ms:.2f}ms: {e}")
```

Ruff (0.12.2)

Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)